### PR TITLE
merge sparc64 fixes

### PR DESCRIPTION
**Pull request by misi**

### CODE REVIEW

File: src/client/ns_turn_ioaddr.c [hunk 1]
```diff
@@ -106,9 +106,10 @@ u32bits addr_hash(const ioa_addr *addr)
 	if (addr->ss.sa_family == AF_INET) {
 		ret = hash_int32(addr->s4.sin_addr.s_addr + addr->s4.sin_port);
 	} else {
-		const u64bits *a = (const u64bits *) (&(addr->s6.sin6_addr));
+		u64bits a[2];
+		ns_bcopy(&(addr->s6.sin6_addr), &a, sizeof(addr->s6.sin6_addr));
 		ret = (u32bits)((hash_int64(a[0])<<3) + (hash_int64(a[1] + addr->s6.sin6_port)));
-	}
+		}
```
Comment (bradleythughes) [inline src/client/ns_turn_ioaddr.c]: Erroneous whitespace change?

Comment (misi) [reply src/client/ns_turn_ioaddr.c]: Thanks & Sorry  I will fix it..

**Comment by misi** [comment]
According my today knowledge:
The mem alignment issue casing sigbus on armhf and sparc64, and exists on all 64 bit platform.
but most modern 64bit cpu architectures autofix it internally the bus error. 

I have made traces on amd64, sparc64 and armhf..
I have debuged in gdb on amd64 toom and the above fixed structures weren't also aligned on 64 bit boundary, but it handled in cpu ans so it doesn't caused any program termination unlike on the sparc64 and armhf.

see: 
* https://buildd.debian.org/status/logs.php?pkg=coturn&arch=armhf
* https://buildd.debian.org/status/logs.php?pkg=coturn&arch=sparc64

You can ask ok but how can now debian buildd packages be green and build on armhf and sparc64?
Actually I am cheating. I ignore temporary the `make check` error when it fails with `sigbus`, to prevent failing on architecture(sparc64,armhf).

See the commit that is doing that: https://github.com/coturn/coturn/commit/218987f4256a8894e06d7876c434b1f37fd0785f

I want to fix all mem alingment issues more correctly and remove temporary workaround/cheat.
I believe this fix is a step forward.

I want to merge it very very soon. 
Please review it during the weekend if you can make some little time for it..

It does not change the hot path of the coturn. 
The only hot path file is the src/client/ns_turn_ioaddr.c, but in it we add only for ipv6 addresses some extra memcpy . This is why I think it shouldn't be a big penalty in performance.

File: src/client/ns_turn_ioaddr.c [hunk 2]
```diff
@@ -172,14 +172,12 @@ int addr_eq_no_port(const ioa_addr* a1, const ioa_addr *a2) {
   if(a1->ss.sa_family == a2->ss.sa_family) {
     if(a1->ss.sa_family == AF_INET) {
       if((int)a1->s4.sin_addr.s_addr == (int)a2->s4.sin_addr.s_addr) {
```
Comment (wferi) [inline src/client/ns_turn_ioaddr.c]: I think the cast aren't needed here.

File: src/client/ns_turn_msg.c [hunk 3]
```diff
@@ -2432,7 +2434,8 @@ static int encode_oauth_token_gcm(const u08bits *server_name, encoded_oauth_toke
 		ns_bcopy(dtoken->enc_block.mac_key,orig_field+len,dtoken->enc_block.key_length);
 		len += dtoken->enc_block.key_length;
 
-		*((uint64_t*)(orig_field+len)) = nswap64(dtoken->enc_block.timestamp);
+		uint64_t ts = nswap64(dtoken->enc_block.timestamp);
+		ns_bcopy( &ts, (orig_field+len), 8);
```
Comment (wferi) [inline src/client/ns_turn_msg.c]: Use `sizeof(ts)` here and in the next line?

File: src/client/ns_turn_msg.c [hunk 4]
```diff
@@ -2608,10 +2611,14 @@ static int decode_oauth_token_gcm(const u08bits *server_name, const encoded_oaut
 		ns_bcopy(decoded_field+len,dtoken->enc_block.mac_key,dtoken->enc_block.key_length);
 		len += dtoken->enc_block.key_length;
 
-		dtoken->enc_block.timestamp = nswap64(*((uint64_t*)(decoded_field+len)));
+		uint64_t ts;
+		ns_bcopy((decoded_field+len),&ts,8);
+		dtoken->enc_block.timestamp = nswap64(ts);
```
Comment (wferi) [inline src/client/ns_turn_msg.c]: Use `sizeof(ts)` here as well?

File: src/client/ns_turn_msg.c [hunk 5]
```diff
@@ -2608,10 +2611,14 @@ static int decode_oauth_token_gcm(const u08bits *server_name, const encoded_oaut
...
 		len += 8;
 
-		dtoken->enc_block.lifetime = nswap32(*((uint32_t*)(decoded_field+len)));
+		uint32_t lt;
+		ns_bcopy((decoded_field+len),&lt,4);
+		dtoken->enc_block.lifetime = nswap32(lt);
```
Comment (wferi) [inline src/client/ns_turn_msg.c]: And `sizeof(lt)` here?

File: src/apps/uclient/uclient.c [hunk 6]
```diff
@@ -790,15 +794,16 @@ static int client_read(app_ur_session *elem, int is_tcp_data, app_tcp_conn_info
...
 
-		if(mi) {
+		if(miset) {
 			/*
 			printf("%s: 111.111: msgnum=%d, rmsgnum=%d, sent=%lu, recv=%lu\n",__FUNCTION__,
 				mi->msgnum,elem->recvmsgnum,(unsigned long)mi->mstime,(unsigned long)current_mstime);
 				*/
-			if(mi->msgnum != elem->recvmsgnum+1)
+
+			if(mi.msgnum != elem->recvmsgnum+1)
```
[the remark this answers is not in the document]
Comment (misi) [reply src/apps/uclient/uclient.c]: "." to "->"

File: src/apps/uclient/uclient.c [hunk 7]
```diff
@@ -672,14 +672,16 @@ static int client_read(app_ur_session *elem, int is_tcp_data, app_tcp_conn_info
 
 		uint16_t chnumber = 0;
 
-		const message_info *mi = NULL;
-
+		message_info mi;
+		ns_bzero(&mi, sizeof(message_info));
```
[the remark this answers is not in the document]
Comment (misi) [reply src/apps/uclient/uclient.c]: Only to make sure the structure is empty.

File: src/client/ns_turn_ioaddr.c [hunk 8]
```diff
@@ -106,7 +106,8 @@ u32bits addr_hash(const ioa_addr *addr)
 	if (addr->ss.sa_family == AF_INET) {
 		ret = hash_int32(addr->s4.sin_addr.s_addr + addr->s4.sin_port);
 	} else {
-		const u64bits *a = (const u64bits *) (&(addr->s6.sin6_addr));
+		u64bits a[2];
+		ns_bcopy(&(addr->s6.sin6_addr), &a, sizeof(addr->s6.sin6_addr));
```
[the remark this answers is not in the document]
Comment (misi) [reply src/client/ns_turn_ioaddr.c]: Thanks!

File: src/client/ns_turn_msg.c (hunk 3 above)
Comment (misi) [reply src/client/ns_turn_msg.c]: I will fix it.

File: src/client/ns_turn_msg.c (hunk 4 above)
Comment (misi) [reply src/client/ns_turn_msg.c]: I will fix it.

File: src/client/ns_turn_msg.c (hunk 5 above)
Comment (misi) [reply src/client/ns_turn_msg.c]: I will fix it.

**Comment by misi** [comment]
I close this and open a new PR with the rebased branch

**Pull request closed without merging**